src/lib.rs

```diff
@@ -64,21 +64,29 @@ pub trait Columnar : 'static {
     /// For example, when comparing two references `Ref<'a>` and `Ref<'b>`, we can
     /// reborrow both to a local lifetime to compare them. This allows us to keep the
     /// lifetimes `'a` and `'b` separate, while otherwise Rust would unify them.
-    fn reborrow<'b, 'a: 'b>(thing: Self::Ref<'a>) -> Self::Ref<'b> where Self: 'a;
+    #[inline(always)] fn reborrow<'b, 'a: 'b>(thing: Self::Ref<'a>) -> Self::Ref<'b> {
+        <Self::Container as Container>::reborrow_ref(thing)
+    }
 }
 
 /// A container that can hold `C`, and provide its preferred references.
 ///
 /// As an example, `(Vec<A>, Vecs<Vec<B>>)`.
-pub trait Container<C: Columnar + ?Sized> {
+pub trait Container : Len + Clear + for<'a> Push<Self::Ref<'a>> + Clone + Default + Send + 'static {
+    /// For each lifetime, a reference with that lifetime.
+    ///
+    /// As an example, `(&'a A, &'a [B])`.
+    type Ref<'a> : Copy;
```

Comment:
Is there a reason to keep `Columnar::Ref` around, other than convenience? By the definition of `Columnar::Container`, the two `Ref` types need to be equal.

Reply:
Convenience for the moment, and to avoid even larger breaks with users who expect it. I'm finding in rewriting some `datatoad` stuff that it is most idiomatic to move away from using this reference and towards using the container reference though. Perhaps a bit more porting and it will be clear, but it will also be annoying to write `<<T as Columnar>::Container as Container>::Ref<'a>`. :D